src/carousel/index.ts

```diff
@@ -0,0 +1,221 @@
+
+    if (!(this.state.transitioning && onEdge)) {
+      // make the transition
+      this.set({ currentSlide: slide, transitioning: true });
+
+      if (onEdge) {
+        // reset to non-cloned slide
+        // tslint:disable-next-line:max-line-length
+        this.animationEndCallback = setTimeout(() => this.resetToRealSlide(from, slide, length), speed);
+      } else {
+        this.refs.track.addEventListener('transitionend', this.disableTransition, false);
+      }
+    }
+  }
+
+  resetToRealSlide = (from: number, to: number, length: number) => {
+    this.state.transitioning = false;
+    this.resetCurrentSlideNum(from, to, length);
+    this.animationEndCallback = null;
+  }
+
+  resetCurrentSlideNum = (from: number, to: number, length: number) => {
+    const currentSlide = from < to ? to % length : (to % length + length) % length;
+    this.set({ currentSlide });
+  }
+
+  disableTransition = () => {
+    this.refs.track.removeEventListener('transitionend', this.disableTransition, false);
+    this.set({ transitioning: false });
+  }
+
+  trackStyle = () => {
+    const { speed } = this.state.settings;
+    if (this.state.transitioning && speed) {
+      const transition = `${speed}ms ease`;
+      const transitionStyles = {
+        '-webkit-transition': transition,
+        transition,
+        '-ms-transition': transition
+      };
+      return { ...this.getStaticTrackStyle(), ...transitionStyles };
+    } else {
+      return this.getStaticTrackStyle();
+    }
+  }
+
+  getStaticTrackStyle = () => {
+    const slideWidth = this.getSlideWidth();
+    const slideCount = this.props.items.length;
+    const slidesToShow = this.state.settings.slidesToShow;
+    const trackWidth = (slideCount + 4 * slidesToShow - 2) * slideWidth;
+    const pos = Carousel.calculatePosition(this.state.currentSlide, slideWidth, slidesToShow);
+    const transform = `translateX(-${pos}px)`;
+    const transformStyles = {
+      transform,
+      '-webkit-transform': transform,
+      '-ms-transform': transform
+    };
+
+    return { width: `${trackWidth}px`, ...transformStyles };
+  }
+
+  getSlideWidth = () => {
+    const visibleWidth = this.refs.wrapper.offsetWidth;
+
+    if (visibleWidth) {
+      return visibleWidth / this.state.settings.slidesToShow;
+    }
+  }
+
+  onTouchStart = (e: TouchEvent & Carousel.Event) => {
+    this.state.touchObject = {
+      startX: e.touches[0].pageX,
+      startY: e.touches[0].pageY
+    };
+
+    e.target.addEventListener('touchend', this.onTouchEnd);
+  }
+
+  onTouchEnd = (e: TouchEvent & Carousel.Event) => {
+    const curX = e.changedTouches[0].pageX;
+    const curY = e.changedTouches[0].pageY;
+    const swipeNext = Carousel.isSwipeToNext({ ...this.state.touchObject }, curX, curY);
+
+    // swipe distance needs to be more than 20
+    if (Math.abs(curX - this.state.touchObject.startX) > MIN_SWIPE_DISTANCE && typeof swipeNext === 'boolean') {
+      swipeNext ? this.moveNext() : this.movePrevious();
```

Comment:
imo I think typeof swipeNext === 'boolean' is unclear

I would rename swipeNext to swipeRight instead and isSwipeToNext to isSwipeToRight

also have isSwipeToRight explicitly return Null in the case we discussed about and do a check for swipeRight !== Null instead

Reply:
we agreed to keep the name as is.
return null has been added. good point